Scripts/Helpers/ProxyUI.cs

```diff
@@ -0,0 +1,549 @@
+
+            // If affordanceMapDefinitions is null, set the list below into the map after setup
+            List<AffordanceDefinition> generatedAffordanceDefinitions = new List<AffordanceDefinition>();
+            var defaultAffordanceVisibilityDefinition = m_AffordanceMap.defaultAffordanceVisibilityDefinition;
+            var defaultAffordanceAnimationDefinition = m_AffordanceMap.defaultAnimationDefinition;
+            foreach (var proxyAffordance in affordances)
+            {
+                var renderers = proxyAffordance.renderer.GetComponentsInChildren<Renderer>(true);
+                if (renderers != null)
+                {
+                    // Setup animated color or alpha transparency for all materials associated with all renderers associated with the control
+                    ProxyAffordanceMap.AffordanceVisibilityDefinition visibilityDefinition;
+                    ProxyAffordanceMap.AffordanceAnimationDefinition animationDefinition;
+                    var control = proxyAffordance.control;
+
+                    // Assemble a new affordance definition and visibility definition for the affordance,
+                    // if a custom definition for the control was not defined in the AffordanceMap
+                    var matchingAffordanceDefinition = affordancesDefinedInMap ? affordanceMapDefinitions.FirstOrDefault(x => x.control == control) : null;
+                    if (matchingAffordanceDefinition == null)
+                    {
+                        // Deep copy the default visibility definition values into a new generated visibility defintion, to be set on a newly generated affordance
+                        visibilityDefinition = new ProxyAffordanceMap.AffordanceVisibilityDefinition
+                        {
+                            visibilityType = defaultAffordanceVisibilityDefinition.visibilityType,
+                            colorProperty = defaultAffordanceVisibilityDefinition.colorProperty,
+                            alphaProperty = defaultAffordanceVisibilityDefinition.alphaProperty,
+                            hiddenColor = defaultAffordanceVisibilityDefinition.hiddenColor,
+                            hiddenAlpha = defaultAffordanceVisibilityDefinition.hiddenAlpha,
+                            hiddenMaterial = defaultAffordanceVisibilityDefinition.hiddenMaterial
+                        };
+
+                        animationDefinition = new ProxyAffordanceMap.AffordanceAnimationDefinition
+                        {
+                            translateAxes = defaultAffordanceAnimationDefinition.translateAxes,
+                            rotateAxes = defaultAffordanceAnimationDefinition.rotateAxes,
+                            min = defaultAffordanceAnimationDefinition.min,
+                            max = defaultAffordanceAnimationDefinition.max,
+                            reverseForRightHand = defaultAffordanceAnimationDefinition.reverseForRightHand
+                        };
```

Comment:
I recommend making AffordanceVisibilityDefinition and AffordanceAnimationDefinition to avoid having to deep copy like this. Generally, I think we'll always want to deep copy these things around.

Reply:
As per our chat, you meant Struct here, right? @mtschoen-unity

Reply:
They've been refactored into classes re a previous PR comment, but I remember our Struct chat.

Reply:
yeah. I was thinking those two types would be structs. That way you don't need to worry about deep copies or shared references

Reply:
If we go this route, we lose the following default field initializers in AffordanceVisibilityDefinition : m_ColorProperty, m_AlphaProperty, and m_HiddenColor.   We also lose the m_Max initialized value in AffordanceAnimationDefinition.  If these are converted to structs, are you suggesting that those values, and all others required for initialization, be moved to a constructor?  Those defaults are nice to have in the inspector when creating a new definition, especially for new users using the standard shader imo.  Thoughts @AndrewTHEManeri  & @mtschoen-unity

Reply:
Yeah, we should be able to handle the default values with the constructor. It's a bit of an annoying rule, but I'm sure there are reasons. I wouldn't say we should get rid of the defaults.

Though, I'm not sure if the serialized fields get those defaults when they're set in the constructor. If that doesn't work, then let's just leave them as classes.

Reply:
How about just making a copy/clone constructor?  If we use these as arguments (I know we do for the visibilitydefinition) they're heavier to copy as arguments than I would want